include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+                       const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+                       void *data);
+
+    /**
+     * A callback function used to unpack the data on the current mesh that
+     * has been packed up previously on the mesh before refinement,
+     * coarsening and repartitioning.
+     */
+    void unpack_function (const typename parallel::distributed::Triangulation<dim,dim>::cell_iterator &cell,
+                          const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+                          const void *data);
+
+    /**
+     * FiniteElement used to project data from and to quadrature points.
+     */
+    const std::shared_ptr<const FiniteElement<dim> > projection_fe;
+
+    /**
+     * Number of DoFs per cell of the projection_fe
+     */
+    const unsigned int dofs_per_cell;
+
+    /**
+     * The size of the data that will be sent, which depends on the DataType class.
+     */
+    size_t size;
+
+    /**
+     * Number of quadrature points at which DataType is stored.
+     */
+    const unsigned int n_q_points;
+
+    /**
+     * Projection matrix from the quadrature points to local DoFs for a single
+     * scalar.
+     */
+    FullMatrix<double> project_to_fe_matrix;
+
+    /**
+     * Projection matrix from the local DoFs to quadrature points for a single
+     * scalar.
+     */
+    FullMatrix<double> project_to_qp_matrix;
+
+    /**
+     * Auxiliary matrix which represents projection of each internal value stored
+     * at the quadrature point (second index) to the local DoFs of the @p projection_fe
+     * (first index).
+     */
+    FullMatrix<double> matrix_dofs;
+
+    /**
+     * Projection of @p matrix_dofs to each child cell in case of adaptive refinement.
+     */
+    FullMatrix<double> matrix_dofs_child;
+
+    /**
+     * Auxiliary matrix which represents data (second index) stored at each
+     * quadrature point (first index).
+     */
+    FullMatrix<double> matrix_quadrature;
+
+    /**
+     * The offset that the Triangulation has assigned to this object
+     * starting at which we are allowed to read.
+     */
+    unsigned int offset;
+
+    /**
+     * A pointer to the CellDataStorage class whose data will be transfered.
+     */
+    CellDataStorage<CellIteratorType,DataType> *data_storage;
+
+    /**
+     * A pointer to the distributed triangulation to which cell data is attached.
+     */
+    parallel::distributed::Triangulation<dim> *tr;
+  };
+
+} // distributed
+
+#endif
+
+/*@}*/
+
+#ifndef DOXYGEN
+
+// -------------------  inline and template functions ----------------
+
+//--------------------------------------------------------------------
+//                         TransferableQuadraturePointData
+//--------------------------------------------------------------------
+TransferableQuadraturePointData::TransferableQuadraturePointData()
+{};
```

Comment:
You'll have to mark this function and the next as `inline` unless you want to get linker errors.